### ISSUE TITLE
We need a way to detect that the connection to sbot is broken

### ISSUE THREAD

**Issue by soapdog** [soapdog/patchfox#29]
Currently there is no way to detect when the connection to _sbot_ is lost. The page just stops. A simple refresh will reestablish it but this currently requires the user to notice and do it. This is a bug,.

- [ ] change online indicator in the avatar when connection is lost.
- [ ] reestablish connection on the background.

**Comment by cryptix** [comment]
I‘m relatively certain there is a reconnecting ssb-client already... at least I saw one once but might have been an old patchwork.. cc @mixmix

other than that I would just do a setInterval thing on sbot.whoami as a kludge?

**Comment by soapdog** [comment]
@cryptix omg that setInterval hack, I am definitely using that I think. I know that losing the connection causes an exception in the console. I just don't know where is the best place to trap it.

**Comment by mixmix** [comment]
This magic
https://github.com/ssbc/patchcore/blob/d2f310a16a8a0b6fed80d8a4afa7cadb21610901/sbot.js#L60

**Comment by soapdog** [comment]
@mixmix that does look like magic, I wish I understood pull-streams better. I am but a newbie regarding regarding them. That code is using depject and patchcore as well. I found it a bit hard to follow how it works, how `pull-reconnect` is supposed to be used. At the moment I've added some mitigation code to keep the connection up, attempting to reconnect once it detects that it is down or if nothing works presenting a nice error screen with options to solve it.

So far it appears to be working but I need to test more. Thanks :heart: again for sending that code to me though. Depending on how it goes, I'll try to reassemble it into something more self contained or minimal in the future.

**Comment by soapdog** [comment]
Of course all the mitigation code just failed in my face while I was working in this message... This is how the error page looks:

![Patchfox - Firefox Nightly 2019-06-15 12 58 11](https://user-images.githubusercontent.com/23247/59551176-53b3a480-8f6d-11e9-8403-30544f7b61b3.png)

At least clicking the `try to reconnect` option always work:

![Patchfox - Firefox Nightly 2019-06-15 12 59 28](https://user-images.githubusercontent.com/23247/59551184-75ad2700-8f6d-11e9-85b5-03c8d69fd665.png)

There is code in the background to attempt reconnection if that exact error is detected but apparently it didn't trigger.

**Comment by mixmix** [comment]
you can ignore all the depject stuff around it. The point is that there's an "observeable" which contains the current connection (either a valid sbot connection, or null). This can be then passed around and you can "watch" that observeable and run things only when it's live

**Comment by soapdog** [comment]
Breaking the connection causes an exception now. This is trapped. I'm trying to find a better way to handle it. I need to investigate `pull-reconnect` and see how to use it. I'm gonna close this issue as it is detectable now even though the fix is not up to the same quality as the UX in patchbay and patchwork.
